### PR TITLE
added support for gradient checkpointing in ESM models

### LINKED ISSUE #24602: Support gradient checkpointing for ESM models
Would you please add `gradient_checkpointing_enable()` feature for ESM models?
These models currently are the best available pre-trained protein language models for researchers.
Many thanks.

### PR DESCRIPTION
**Pull request by sanjeevk-os**
# What does this PR do?

Fixes #24602

Adds gradient checkpointing support for ESM models

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
    Issue link: https://github.com/huggingface/transformers/issues/24602
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@Rocketknight1 @amyeroberts : Please review and suggest any required changes, thanks.

### CODE REVIEW

**Review by Rocketknight1 — APPROVED** [review]
Looks good to me, and the test looks clean too!

Amy is out but pinging @LysandreJik for core maintainer review.

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @sanjeevk-os!
